DOC: Fix description of dtype default in linspace

bjnath opened the pull request:
Clarify linspace docstring to say the default dtype will never be smaller than float.

Fixes gh-8597

bjnath committed:
DOC: Fix description of dtype default in linspace

Fixes gh-8597

rossbar reviewed (commented):
I don't think "smaller than float" is an accurate wording - from the discussion in #8597, the main takeaway is that the inferred type will be a floating point type  in the case where the inputs (`start`, `stop`, and `step`) are all integer types (this makes sense given what linspace does, but the wording for the dtype parameter description is not very clear on this point). 

Whatever wording is chosen should also accommodate complex `start`/`stop` values as well.

bjnath commented:
I think the new wording says what you believe it needs to. "Infers" is from the original and seems like the right approach;  readers can understand this means the dtype chosen will handle complex arguments correctly. But "infers" also suggests it would choose an integer for integer start and stop, and "smaller than float" indicates it won't.

rossbar commented:
> "Infers" is from the original and seems like the right approach;

Agreed! No problems at all with this wording.

> But "infers" also suggests it would choose an integer for integer start and stop,

Agreed here as well - this is what I understand to be the core of #8597

> "smaller than float" indicates it won't.

I wasn't very clear in my original comment, let me try to rephrase: I think the specific wording "smaller than float" is imprecise and I had trouble interpreting it; i.e. in what sense is a floating point type smaller than an integer type? It's not clear to me how "smaller" maps onto the discussion about types. It could be that I'm just missing something.

Re: my comment about complex numbers, I was mostly referring to this comment from #8597:

> Don't think its wrong, but we can add "a floating point type inferred" or so.

I was just thinking that we should be careful about wording like this as well, because it might introduce unintended ambiguities with respect to complex types. As I understand the [NumPy type hierarchy](https://numpy.org/devdocs/reference/arrays.scalars.html), complex types are distinct from floating point types, so saying "a floating point type is inferred" would technically be incorrect in the case that `start`/`stop` are complex.

bjnath commented:
Ah, we're closer than I'd thought. I see your point about "smaller," too.  I was thinking "smaller" in the sense of the promotion hierarchy. But I failed to make it clear, so I'll revisit.

>  we should be careful about wording like this as well

My thought too, same reason.

bjnath committed:
DOC: Redrafted fix for description of dtype default in linspace

Replaces 'smaller than' with more explicit wording.

bjnath commented:
2db208c rewords to 

> If `dtype` is not given, the data type is inferred from `start`, `stop`, and `num`. The inferred dtype will never be an integer; `float` is chosen even if all three arguments are  integers.

rossbar commented on numpy/core/function_base.py:
```diff
@@ -52,8 +52,10 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
-        The type of the output array.  If `dtype` is not given, infer the data
-        type from the other input arguments.
+        The type of the output array.  If `dtype` is not given, the data type
+        is inferred from `start`, `stop`, and `num`. The inferred dtype will
```
I would leave `num` off the list - it must be an integer, but for the purposes of determining the dtype it is cast to float internally.

bjnath replied:
Agreed, `num` is dont-care.

rossbar reviewed (commented):
This is an improvement over the last iteration and essentially addresses #8597. I'm not sure a change in the documentation is necessarily an improvement as I think the original wording (combined with an understanding of what linspace does) should be sufficient to convince users that not to expect an integer type from linspace by default. 

It would be nice to be able to say "infer the appropriate inexact type from the data", but I don't think that `np.inexact` is necessarily well-known enough to clarify things for most users. 

In conclusion - Let's see what other reviewers think! If the consensus is that it's an improvement to explicitly say that `linspace` won't return integer arrays by default, then the wording here accomplishes that.

Qiyu8 commented:
I found three more "infer the data type from the other input arguments" statement which should clarified as well.

- line 1324 from `_add_newdocs.py`
- line 205 and line 300 from `function_base.py`

bjnath commented:
> not to expect an integer type from linspace
@rossbar, why would I not expect th dtype here to have been integer?

> In [2]: np.linspace(start = 0, stop = 100, num = 5)
> Out[2]: array([  0.,  25.,  50.,  75., 100.])
>

bjnath committed:
DOC: Drop 'num' from inference rules in PR #16433

Per https://github.com/numpy/numpy/pull/16433#discussion_r434295779,
now says the inferred dtype depends only on 'start' and 'stop'.

bjnath commented:
Commit 5487847 resolves https://github.com/numpy/numpy/pull/16433#discussion_r434295779.
Text is now

>     dtype : dtype, optional
>         The type of the output array.  If `dtype` is not given, the data type
>         is inferred from `start` and `stop`. The inferred dtype will
>         never be an integer; `float` is chosen even if the arguments would
>         produce an array of integers.
> 

@Qiyu8, thanks for finding these:
> I found three more "infer the data type from the other input arguments" statement which should clarified as well.

I'll revise the PR to include them once new wording is chosen.

rossbar commented:
> @rossbar, why would I not expect the dtype here to have been integer?

Due to the fact that linspace evenly subdivides an interval. The fact that np.inexact types are returned is consistent with Python itself (since Py3k anyways) in that true division is the default, regardless of input types.

Thanks for the `grep` @Qiyu8 - the fact that there are more instances makes me more inclined to keep the current wording. Let's see if any other reviewers can weigh in!

bjnath commented:
> Due to the fact that linspace evenly subdivides an interval. The fact that np.inexact types are returned is consistent with Python itself (since Py3k anyways) in that true division is the default, regardless of input types.

I think you've made my case. It took 35 words to explain why the behavior is obvious. 

Therefore, to many users, it *won't* be obvious. Would it kill us to spell it out? 

> the fact that there are more instances makes me more inclined to keep the current wording.

What's the reasoning here?

bjnath commented:
Also, linspace is one of the first NumPy commands shown to beginners. 

However obvious the float outcome is to an experienced user, it's not going to be obvious to them.

How much testimony do we need that the default behavior confuses people? It confused me, it confused whoever raised gh-8597, and it confused [this unfortunate blogger](https://www.sharpsightlabs.com/blog/numpy-linspace/), who thought the result of
```
np.linspace(start = 0, stop = 100, num = 5)
```
was so obviously integer that they didn't bother running  it first to check:
![image](https://user-images.githubusercontent.com/6691888/83699369-d647d080-a5d1-11ea-9d94-aa9f92b24f4e.png)

Why so much hesitation over this?  What's at stake here that I'm missing?

rgommers commented:
The addition of 
```
The inferred dtype will never be an integer; `float` is chosen even if the
arguments would produce an array of integers.
```
seems clear to me. Happy to go with that.

Agree with "inexact" not being useful here to clarify, too obscure.

bjnath commented:
Thanks, @rossbar . How about those other instances that @Qiyu8 found -- do they still need to be looked at?

rossbar commented:
They certainly can be if you think a wording update would improve clarity!
